airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/component/TableSchemaEvolutionClient.kt

```diff
@@ -57,5 +57,142 @@ interface TableSchemaEvolutionClient {
+            tableName,
+            expectedSchema,
+            expectedAdditionalInfo,
+            columnChangeset,
+            additionalChangeset
+        )
+    }
+
+    /**
+     * Query the destination and discover the schema of an existing table. If this method includes
+     * the `_airbyte_*` columns, then [computeSchema] MUST also include those columns.
+     *
+     * The `Any?` return value should be used for anything not represented in the [TableSchema]
+     * object. For example, you may want to configure a partitioning/clustering key on the table,
+     * based on the sync mode - so you would want to return the existing table's partition/cluster
+     * keys.
+     *
+     * Most destinations will likely just use `null` for that value.
+     */
+    suspend fun discoverSchema(tableName: TableName): Pair<TableSchema, Any?>
+
+    /**
+     * Compute the schema that we _expect_ the table to have, given the [stream]. This should _not_
+     * query the destination in any way.
+     */
+    fun computeSchema(
+        stream: DestinationStream,
+        columnNameMapping: ColumnNameMapping
+    ): Pair<TableSchema, Any?>
+
+    /**
+     * This function computes a changeset between two "additional info" structs. This is the `Any?`
+     * value returned from [discoverSchema].
+     *
+     * If your destination doesn't return anything interesting in that field, you should just
+     * `return null` from this function.
+     */
+    fun computeAdditionalChangeset(actualSchemaInfo: Any?, expectedSchemaInfo: Any?): Any?
+
+    /**
+     * Execute the changeset against the destination. After this method completes, a call to
+     * [discoverSchema] should return an identical schema as [computeSchema].
+     */
+    suspend fun applyChangeset(
+        // Eventually it would be nice for the stream+columnnamemapping to go away,
+        // but that would require computeSchema() to include the airbyte columns,
+        // and we're not consistent about doing that (and there's some CDK work needed
+        // to make that easier to do anyway).
+        // So for now just include the full stream object.
+        // This is needed for destinations that need to recreate the entire table.
+        stream: DestinationStream,
+        columnNameMapping: ColumnNameMapping,
+        tableName: TableName,
+        expectedSchema: TableSchema,
+        expectedAdditionalInfo: Any?,
+        columnChangeset: ColumnChangeset,
+        additionalSchemaInfoChangeset: Any?,
     )
 }
+
+data class TableSchema(
+    /**
+     * A map from column name to type. Note that the column name should be as it appears in the
+     * destination: for example, Snowflake upcases all identifiers, so these column names should be
+     * upcased.
+     */
+    val columns: Map<String, ColumnType>
+) {
+    /** Generate a changeset which, when applied to `this`, will result in [expectedSchema]. */
+    fun computeChangeset(expectedSchema: TableSchema): ColumnChangeset {
```

Comment:
Maybe move this to the interface in-case someone needs to override it / we want to mock it.